Readme.md
<div align="center">
  <img src="./docs/assets/logo.png" alt="Postcodes.js logo">

  <p>Postcodes.js is an API wrapper for postcodes.io allowing users to find information about UK postcodes, find UK postcodes from co-ordinates and more for free, with no API key required and with no rate-limits. Postcodes.js features caching to create a fast, efficient library for UK postcode lookup.</p>

  <a href="https://npmjs.com/package/postcodes.js">
    <img src="https://img.shields.io/npm/v/postcodes.js">
  </a>

  <a href="./LICENSE">
    <img src="https://img.shields.io/badge/license-MIT-blue">
  </a>
</div>

<br>

> Example: Find data about the postcode "SW1W 0DT" and output the latitude and longitude.

```js
Postcodes.lookup("SW1W 0DT").then(result => {
  // outputs "51.494853 -0.145828"
  console.log(result.latitude, result.longitude);
});
```

## Design

Postcodes.js is designed with efficiency and speed at its heart. Postcode.js is an API wrapper for the [Postcodes.io](https://postcodes.io/) API, which is a free API for finding information and performing other operations on UK postcodes. By default, Postcodes.js uses the public API provided at [api.postcodes.io](https://api.postcodes.io), however, due to Postcodes.io being open source, it's possible to host your own instance and configure Postcodes.js to use that instance. Postcodes.js, by default, uses a cache shared between all of it's instances that caches up to 10,000 results for a maximum of 30 days, greatly improving speed for results already cached and decreasing the amount of API calls needed. It is possible to disable caching all together or provide your own cache to Postcodes.js.

## Basic Usage

You can install Postcodes.js through NPM:

```
$ npm i postcodes.js
```

You can then include the library in Node like:

```js
const PostcodesJS = require("postcodes.js");
const Postcodes = new PostcodesJS();
```

You can then perform operations such as postcode data lookup:

```js
Postcodes.lookup("SW1W 0DT").then(result => {
  console.log(result);
});
```

## Example Usages

Use callbacks instead of promises:

```js
const PostcodesJS = require("postcodes.js");
const Postcodes = new PostcodesJS.Callbacks();

Postcodes.lookup("SW1W 0DT", function(error, result) {
  console.log(result);
});
```

Lookup postcode data:

```js
Postcodes.lookup("SW1W 0DT").then(result => {
  console.log(result);
});
```

Bulk lookup postcodes:

```js
Postcodes.bulkLookup(["SW1W 0DT", "SW1A 2AA"]).then(result => {
  console.log(result);
});
```

Lookup postcodes by co-ordinates:

```js
Postcodes.coords({ longitude: -0.145828, latitude: 51.494853 }).then(result => {
  console.log(result);
});
```

Bulk lookup postcodes by mutliple co-ordinates:

```js
Postcodes.bulkCoords([
  { latitude: 51.494853, longitude: -0.145828 },
  { latitude: 51.503297, longitude: -0.127882 }
]).then(result => {
  console.log(result);
});
```

Get random postcode:

```js
Postcodes.random().then(result => {
  console.log(result);
});
```

Validate a postcode is a valid postcode:

```js
Postcodes.validate("WRONG POSTCODE").then(result => {
  // outputs false
  console.log(result);
});
```

Autocomplete start of a postcode:

```js
Postcodes.autocomplete("SW1A").then(result => {
  console.log(result);
});
```

Query postcode information:

```js
Postcodes.query("SW1W 0DT").then(result => {
  console.log(result);
});
```

Get data about a terminated postcode (a postcode that is no longer in use):

```js
Postcodes.terminated("E1W 1UU").then(result => {
  console.log(result);
});
```

Get data about an outcode:

```js
Postcodes.outcode("SW1W").then(result => {
  console.log(result);
});
```

Get nearest outcodes from an outcode:

```js
Postcodes.nearestOutcode("SW1W").then(result => {
  console.log(result);
});
```

Get outcode from co-ordinates:

```js
Postcodes.outcodeFromCoord({ longitude: -0.145828, latitude: 51.494853 }).then(
  result => {
    console.log(result);
  }
);
```

Use different Postcodes.io host URL:

```js
const Postcodes = new PostcodesJS({
  host: "postcodesio.example.com"
});
```

Disable caching:

```js
const Postcodes = new PostcodesJS({
  cache: false
});
```

Provide custom cache:

```js
const LRUCache = require("lru-cache");
const Postcodes = new PostcodesJS({
  cacheInstance: new LRUCache({
    max: 10000,
    maxAge: 90 * 24 * 60 * 60 * 1000
  })
});
```

Use HTTP (unsecure) connection instead of default HTTPS (secure) connection:

```js
const Postcodes = new PostcodesJS({
  https: false
});
```

Change hashing algorithm used for hashing queries to be saved in the cache (defaults to "sha256"). Can provide any hashing algorithm name that can be used with `crypto.createHash(hashingAlgorithm)`.

```js
const Postcodes = new PostcodesJS({
  hashingAlgorithm: "md5"
});
```

Provide a useragent to be used with requests to the API:

```js
const Postcodes = new PostcodesJS({
  useragent: "Custom Useragent"
});
```

## Credit

Author: [Tom](https://github.com/TomPrograms)

## License

[MIT](./LICENSE)


Changelog.md
# 1.0.3 - 02/02/2021

### Updated

- Dependencies

# 1.0.2 - 24/04/2020

### Fixed

- Author's name in package.json, license and Readme.md.

# 1.0.1 - 17/03/2020

### Removed

- Unnecessary `console.log` statement.

# 1.0.0 - 17/03/2020

- 🎉 Initial Release of Postcodes.js


src/postcodes.js
const LRUCache = require("lru-cache");
const axios = require("axios");
const crypto = require("crypto");

const globalCache = new LRUCache({
  max: 10000, // maximum 10,000 items
  maxAge: 30 * 24 * 60 * 60 * 1000 // maximum age of 30 days
});

class Request {
  constructor(requestData) {
    this.requestData = requestData;
  }

  createHash(hashingAlgorithm) {
    this.hash = crypto
      .createHash(hashingAlgorithm)
      .update(JSON.stringify(this.requestData))
      .digest("hex");
  }
}

class PostcodesJS {
  constructor(options = {}) {
    this.cache = options.cache || true;
    this.cacheInstance = options.cacheInstance;
    this.host = options.host || "api.postcodes.io";
    this.https = options.https || true;
    this.hashingAlgorithm = options.hashingAlgorithm || "sha256";
    this.useragent = options.useragent
      ? `${options.useragent} powered by Postcodes.js`
      : "Application Powered by Postcodes.js";
  }

  setCache(key, data) {
    (this.cacheInstance || globalCache).set(key, data);
  }

  getCache(key) {
    return (this.cacheInstance || globalCache).get(key);
  }

  getURL(path) {
    let constructed = this.host + path;

    // if no protocol, append appropriate protocol
    let noProtocol =
      !constructed.startsWith("https://") && !constructed.startsWith("http://");
    if (noProtocol) {
      constructed = (this.https ? "https://" : "http://") + constructed;
    }

    // remove any duplicate slashes
    return constructed.replace(/([^:]\/)\/+/g, "$1");
  }

  request(request) {
    if (this.cache === true) {
      let cachedData = this.getCache(request.hash);
      if (cachedData) return Promise.resolve(cachedData);
    }

    const requestData = request.requestData;
    return new Promise((resolve, reject) => {
      axios({
        method: requestData.method,
        url: requestData.url,
        data: requestData.body,
        params: requestData.params,
        headers: {
          "User-Agent": this.useragent
        }
      })
        .then(response => {
          if (this.cache === true) {
            this.setCache(request.hash, response.data.result);
          }

          resolve(response.data.result);
        })
        .catch(error => {
          reject(error);
        });
    });
  }

  constructRequest(url, method = "GET", body, params) {
    const request = new Request({
      url,
      method,
      body,
      params
    });
    request.createHash(this.hashingAlgorithm);

    return this.request(request);
  }

  lookup(postcode) {
    const url = this.getURL(`/postcodes/${postcode}`);
    return this.constructRequest(url);
  }

  bulkLookup(postcodes) {
    const url = this.getURL("/postcodes");
    const body = {
      postcodes
    };
    return this.constructRequest(url, "POST", body);
  }

  coords(coordinate) {
    const url = this.getURL("/postcodes");
    const params = {
      lon: coordinate.longitude,
      lat: coordinate.latitude
    };
    return this.constructRequest(url, "GET", undefined, params);
  }

  bulkCoords(coordinates) {
    const url = this.getURL("/postcodes");
    const body = {
      geolocations: coordinates
    };
    return this.constructRequest(url, "POST", body);
  }

  random() {
    const url = this.getURL("/random/postcodes");
    return this.constructRequest(url);
  }

  validate(postcode) {
    const url = this.getURL(`/postcodes/${postcode}/validate`);
    return this.constructRequest(url);
  }

  nearestPostcode(postcode) {
    const url = this.getURL(`/postcodes/${postcode}/nearest`);
    return this.constructRequest(url);
  }

  autocomplete(postcode) {
    const url = this.getURL(`/postcodes/${postcode}/autocomplete`);
    return this.constructRequest(url);
  }

  query(postcode) {
    const url = this.getURL("/postcodes");
    const params = {
      q: postcode
    };
    return this.constructRequest(url, "GET", undefined, params);
  }

  terminated(postcode) {
    const url = this.getURL(`/terminated_postcodes/${postcode}`);
    return this.constructRequest(url);
  }

  outcode(outcode) {
    const url = this.getURL(`/outcodes/${outcode}`);
    return this.constructRequest(url);
  }

  nearestOutcode(outcode) {
    const url = this.getURL(`/outcodes/${outcode}/nearest`);
    return this.constructRequest(url);
  }

  outcodeFromCoord(coordinate) {
    const url = this.getURL("/outcodes");
    const params = {
      lon: coordinate.longitude,
      lat: coordinate.latitude
    };
    return this.constructRequest(url, "GET", undefined, params);
  }
}

class PostcodesJSCallbacks extends PostcodesJS {
  lookup(postcode, callback) {
    const returnedPromise = super.lookup(postcode);
    this.convertToCallback(returnedPromise, callback);
  }

  bulkLookup(postcodes, callback) {
    const returnedPromise = super.bulkLookup(postcodes);
    this.convertToCallback(returnedPromise, callback);
  }

  coords(coordinate, callback) {
    const returnedPromise = super.coords(coordinate);
    this.convertToCallback(returnedPromise, callback);
  }

  bulkCoords(coordinates, callback) {
    const returnedPromise = super.bulkCoords(coordinates);
    this.convertToCallback(returnedPromise, callback);
  }

  random(callback) {
    const returnedPromise = super.random();
    this.convertToCallback(returnedPromise, callback);
  }

  validate(postcode, callback) {
    const returnedPromise = super.validate(postcode);
    this.convertToCallback(returnedPromise, callback);
  }

  nearestPostcode(postcode, callback) {
    const returnedPromise = super.nearestPostcode(postcode);
    this.convertToCallback(returnedPromise, callback);
  }

  autocomplete(postcode, callback) {
    const returnedPromise = super.autocomplete(postcode);
    this.convertToCallback(returnedPromise, callback);
  }

  query(postcode, callback) {
    const returnedPromise = super.query(postcode);
    this.convertToCallback(returnedPromise, callback);
  }

  terminated(postcode, callback) {
    const returnedPromise = super.terminated(postcode);
    this.convertToCallback(returnedPromise, callback);
  }

  outcode(outcode, callback) {
    const returnedPromise = super.outcode(outcode);
    this.convertToCallback(returnedPromise, callback);
  }

  nearestOutcode(outcode, callback) {
    const returnedPromise = super.nearestOutcode(outcode);
    this.convertToCallback(returnedPromise, callback);
  }

  outcodeFromCoord(coordinate, callback) {
    const returnedPromise = super.outcodeFromCoord(coordinate);
    this.convertToCallback(returnedPromise, callback);
  }

  convertToCallback(promise, callback) {
    if (typeof callback !== "function") return promise;

    return promise
      .then(result => callback(null, result))
      .catch(error => callback(error, null));
  }
}

module.exports = PostcodesJS;
module.exports.Callbacks = PostcodesJSCallbacks;
